cvmfs/sqlitemem.cc

```diff
@@ -0,0 +1,489 @@
+  assert(block_ctl->size <= 0);
+
+  int32_t new_size = -(block_ctl->size);
+  assert(new_size > 0);
+  AvailBlockCtl *new_avail = reinterpret_cast<AvailBlockCtl *>(block_ctl);
+
+  if (prior_tag == kTagAvail) {
+    // Merge with block before and remove the block from the list
+    int32_t prior_size = reinterpret_cast<AvailBlockTag *>(
+      reinterpret_cast<char *>(block_ctl) - sizeof(AvailBlockTag))->size;
+    assert(prior_size > 0);
+    new_size += prior_size;
+    new_avail = reinterpret_cast<AvailBlockCtl *>(
+      reinterpret_cast<char *>(block_ctl) - prior_size);
+    AvailBlockCtl *next = new_avail->GetNextPtr(arena_);
+    AvailBlockCtl *prev = new_avail->GetPrevPtr(arena_);
+    prev->link_next = new_avail->link_next;
+    next->link_prev = new_avail->link_prev;
+    if (rover_ == new_avail)
+      rover_ = head_avail_;
+  }
+
+  int32_t succ_size = *reinterpret_cast<int32_t *>(
+    reinterpret_cast<char *>(new_avail) + new_size);
+  if (succ_size >= 0) {
+    // Merge with succeeding block and remove the block from the list
+    AvailBlockCtl *succ_avail = reinterpret_cast<AvailBlockCtl *>(
+      reinterpret_cast<char *>(new_avail) + new_size);
+    AvailBlockCtl *next = succ_avail->GetNextPtr(arena_);
+    AvailBlockCtl *prev = succ_avail->GetPrevPtr(arena_);
+    prev->link_next = succ_avail->link_next;
+    next->link_prev = succ_avail->link_prev;
+    new_size += succ_size;
+    if (rover_ == succ_avail)
+      rover_ = head_avail_;
+  }
+
+  // Insert new free block at the end of the list
+  AvailBlockCtl *next = head_avail_;
+  AvailBlockCtl *prev = head_avail_->GetPrevPtr(arena_);
+  next->link_prev = new_avail->ConvertToLink(arena_);
+  prev->link_next = new_avail->ConvertToLink(arena_);
+  new_avail->link_next = head_avail_->ConvertToLink(arena_);
+  new_avail->link_prev = prev->ConvertToLink(arena_);
+
+  // Set new free block's boundaries
+  new_avail->size = new_size;
+  void *upper_tag =
+    reinterpret_cast<char *>(new_avail) + new_size - sizeof(AvailBlockTag);
+  new (upper_tag) AvailBlockTag(new_size);
+}
+
+
+/**
+ * Walks the list of available blocks starting from rover and allocates the
+ * first available spot that's large enough.  Puts the reserved block at the end
+ * of the available one and, if necessary, removes the available one from the
+ * list of free blocks.
+ */
+void *MemoryManager::MallocArena::Malloc(const uint32_t size) {
+  assert(size > 0);
+
+  // Control word first, block type tag last
+  int32_t total_size = sizeof(ReservedBlockCtl) + size + 1;
+  if (total_size < kMinBlockSize)
+    total_size = kMinBlockSize;
+
+  bool wrapped = false;
+  // Generally: p = LINK(q)
+  AvailBlockCtl *q = rover_;
+  AvailBlockCtl *p;
+  do {
+    p = q->GetNextPtr(arena_);
+    if (p->size >= total_size)
+      break;
+    if (p == head_avail_) {
+      if (wrapped)
+        return NULL;
+      wrapped = true;
+    }
+    q = p;
+  } while (true);
```

Comment:
I would put that loop into a method `FindFreeBlock()` returning either the block or NULL